execution/function/operator.go

```diff
@@ -33,6 +33,96 @@ type functionOperator struct {
 	pointBuf     []promql.Point
 }
 
+type absentOperator struct {
+	mint        int64
+	maxt        int64
+	step        int64
+	currentStep int64
+	stepsBatch  int
+	funcExpr    *parser.Call
+	series      []labels.Labels
+	once        sync.Once
+
+	next       model.VectorOperator
+	vectorPool *model.VectorPool
+}
+
+func (o *absentOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*absentOperator] %v()", o.funcExpr.Func.Name), []model.VectorOperator{}
+}
+
+func (o *absentOperator) Series(_ context.Context) ([]labels.Labels, error) {
+	o.loadSeries()
+
+	return o.series, nil
+}
+
+func (o *absentOperator) loadSeries() {
+	o.once.Do(func() {
+		// https://github.com/prometheus/prometheus/blob/main/promql/functions.go#L1385
+		var lm []*labels.Matcher
+		switch n := o.funcExpr.Args[0].(type) {
+		case *parser.VectorSelector:
+			lm = n.LabelMatchers
+		case *parser.MatrixSelector:
+			lm = n.VectorSelector.(*parser.VectorSelector).LabelMatchers
+		default:
+			o.series = []labels.Labels{labels.EmptyLabels()}
+			return
+		}
+
+		has := make(map[string]bool)
+		lmap := make(map[string]string)
+		for _, l := range lm {
+			if l.Name == labels.MetricName {
+				continue
+			}
+			if l.Type == labels.MatchEqual && !has[l.Name] {
+				lmap[l.Name] = l.Value
+				has[l.Name] = true
+			} else {
+				delete(lmap, l.Name)
+			}
+		}
+		o.series = []labels.Labels{labels.FromMap(lmap)}
+	})
+}
+
+func (o *absentOperator) GetPool() *model.VectorPool {
+	return o.vectorPool
+}
+
+func (o *absentOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	if o.currentStep > o.maxt {
+		return nil, nil
+	}
+
+	vectors, err := o.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(vectors) > 0 {
+		return nil, nil
+	}
```

Comment:
We probably can optimize a little bit for `o.next.Next(ctx)` call as we don't need to retrieve the whole vectors back, just need to know whether `len(vectors) > 0` or not. But that's not something blocking this pr.

Reply:
My mistake, we probably need whole vector, i added test that fails if we do a range query and some data is there.

Reply:
weirdly enough with latest changes performance has improved drastically

```
RangeQuery/absent_and_exists-8                              132.40m ± ∞ ¹   14.03m ± ∞ ¹  -89.40% (p=0.008 n=5)
RangeQuery/absent_and_doesnt_exist-8                         184.4µ ± ∞ ¹   268.6µ ± ∞ ¹  +45.67% (p=0.008 n=5)
```